coderd/notifications/manager.go

```diff
@@ -0,0 +1,450 @@
+func defaultHandlers(cfg codersdk.NotificationsConfig, log slog.Logger) *HandlerRegistry {
+	reg, err := NewHandlerRegistry(
+		dispatch.NewSMTPHandler(cfg.SMTP, log.Named("dispatcher.smtp")),
+		dispatch.NewWebhookHandler(cfg.Webhook, log.Named("dispatcher.webhook")),
+	)
+	if err != nil {
+		panic(err)
+	}
+	return reg
+}
+
+// WithHandlers allows for tests to inject their own handlers to verify functionality.
+func (m *Manager) WithHandlers(reg *HandlerRegistry) {
+	m.handlers = reg
+}
+
+// Run initiates the control loop in the background, which spawns a given number of notifier goroutines.
+// Manager requires system-level permissions to interact with the store.
+func (m *Manager) Run(ctx context.Context, notifiers int) {
+	// Closes when Stop() is called or context is canceled.
+	go func() {
+		err := m.loop(ctx, notifiers)
+		if err != nil {
+			m.log.Error(ctx, "notification manager stopped with error", slog.Error(err))
+		}
+	}()
+}
+
+// loop contains the main business logic of the notification manager. It is responsible for subscribing to notification
+// events, creating notifiers, and publishing bulk dispatch result updates to the store.
+func (m *Manager) loop(ctx context.Context, notifiers int) error {
+	defer func() {
+		close(m.done)
+		m.log.Info(context.Background(), "notification manager stopped")
+	}()
+
+	// Caught a terminal signal before notifiers were created, exit immediately.
+	select {
+	case <-m.stop:
+		m.log.Warn(ctx, "gracefully stopped")
+		return xerrors.Errorf("gracefully stopped")
+	case <-ctx.Done():
+		m.log.Error(ctx, "ungracefully stopped", slog.Error(ctx.Err()))
+		return xerrors.Errorf("notifications: %w", ctx.Err())
+	default:
+	}
+
+	var (
+		// Buffer successful/failed notification dispatches in memory to reduce load on the store.
+		//
+		// We keep separate buffered for success/failure right now because the bulk updates are already a bit janky,
+		// see BulkMarkNotificationMessagesSent/BulkMarkNotificationMessagesFailed. If we had the ability to batch updates,
+		// like is offered in https://docs.sqlc.dev/en/stable/reference/query-annotations.html#batchmany, we'd have a cleaner
+		// approach to this - but for now this will work fine.
+		success = make(chan dispatchResult, m.cfg.StoreSyncBufferSize)
+		failure = make(chan dispatchResult, m.cfg.StoreSyncBufferSize)
+	)
+
+	// Create a specific number of notifiers to run concurrently.
+	var eg errgroup.Group
+	for i := 0; i < notifiers; i++ {
+		eg.Go(func() error {
+			m.notifierMu.Lock()
+			n := newNotifier(ctx, m.cfg, i+1, m.log, m.store, m.handlers)
+			m.notifiers = append(m.notifiers, n)
+			m.notifierMu.Unlock()
+			return n.run(ctx, success, failure)
+		})
+	}
+
+	eg.Go(func() error {
+		// Every interval, collect the messages in the channels and bulk update them in the database.
+		tick := time.NewTicker(m.cfg.StoreSyncInterval.Value())
```

Comment:
follow-up: I think this would be an excellent place to use our fancy new clock library https://github.com/coder/coder/tree/main/clock.

Reply:
Cool idea, I'll look into it.